Use has_configuration_set() during microsite initialization

felipemontoya opened the pull request:
This PR:

Changes the `microsite_enable_pre_startup` method to use the `has_configuration_set` method to determine whether the path modifications should be made.

Before the change, the logic was tied to the settings file configuration, and could not be modified at the database backend without overriding the complete enable_pre method. After the change, both backends work well with the same enable_pre method.

edx-webhook commented:
Thanks for the pull request, @felipemontoya! I've created [OSPR-1137](https://openedx.atlassian.net/browse/OSPR-1137) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

felipemontoya commented:
@mattdrayer, @ziafazal. This is something minor I found during our internal upgrade to dogwood

mattdrayer commented:
@felipemontoya is it possible to include a test validating this change?

felipemontoya commented:
Hi @mattdrayer, sorry for the late answer, I'll add a test for this the moment we are done upgrading to dogwood, should take me a few days.

mattdrayer commented:
@felipemontoya -- any chance you might be getting around to this PR this week?

felipemontoya commented:
@mattdrayer This was well timed. I actually started working on this a while ago. I will push the tests in a little while

mattdrayer commented:
Great!

felipemontoya commented:
@mattdrayer, There was a test already which tested the behavior of the same method whether the USE_MICROSITES flag was on, I moved that into the test_microsites module and made a new one testing the behavior when the has_configuration_set returned true or false.

felipemontoya commented:
jenkins run bokchoy

mattdrayer commented:
LGTM -- :+1: -- @saleem-latif, can you please review?

saleem-latif commented on common/djangoapps/microsite_configuration/tests/backends/test_base.py:
```diff
@@ -130,3 +138,36 @@ def test_not_yet_implemented(self):
+            if path != settings.MICROSITE_ROOT_DIR
+        ]
+
+        with patch('microsite_configuration.backends.base.BaseMicrositeBackend.has_configuration_set',
```
@felipemontoya I think it would be better if instead of patching `microsite_configuration.backends.base.BaseMicrositeBackend.has_configuration_set ` we patch `settings.MICROSITE_CONFIGURATION` and perform subsequent checks.

felipemontoya replied:
@saleem-latif, wouldn't that defeat the purpose of the PR? The idea is precisely not to use settings.MICROSITE_CONFIGURATION, because it would asume that the backend has to use the settings file, as the DatabaseBackend does not use the settings this PR shifts the responsability to the backend self by using the Backend.has_configuration_set method.

saleem-latif replied:
@felipemontoya my concern here is that we are only testing the situation when `has_configuration_set` returns `True`  or `False`. However, I think it would a good idea to test the following user cases,
1. User sets/leaves-empty the value of `settings.MICROSITE_CONFIGURATION ` 
2. User populates/leaves-empty configuration in `Microsite` table.

here we can patch `settings.MICROSITE_CONFIGURATION ` and in the tests for `microsite_configuration.backends.database.DatabaseMicrositeBackend` we can add/remove configuration in `Microsite` table. 
Maybe we can leave the above test case as it is and add new ones for above user cases.

felipemontoya replied:
I agree that those other cases also need to be tested. And I would vouch for the scenario in which we leave this unit test as it is, testing only the outcome of this method, and we also add new tests for the `has_configuration_set` both here and in the database backend.

mattdrayer replied:
I'm fine leaving this test as-is and adding new tests for those specific cases -- certainly can't hurt to have *more* tests :)

saleem-latif commented on common/djangoapps/microsite_configuration/tests/backends/test_base.py:
```diff
@@ -130,3 +138,36 @@ def test_not_yet_implemented(self):
+            self.assertNotIn(settings.MICROSITE_ROOT_DIR,
+                             settings.DEFAULT_TEMPLATE_ENGINE['DIRS'])
+        with patch('microsite_configuration.backends.base.BaseMicrositeBackend.has_configuration_set',
+                   return_value=True):
```
same as above, it would be better to patch `settings.MICROSITE_CONFIGURATION` instead of `microsite_configuration.backends.base.BaseMicrositeBackend.has_configuration_set `  and perform subsequent checks.

felipemontoya replied:
same as above

saleem-latif commented:
@felipemontoya just a couple of small changes, other than that LGTM :+1:

mattdrayer commented:
@felipemontoya -- just circling back around to this PR -- what are your thoughts on covering the use cases suggested by @saleem-latif?

felipemontoya commented:
@mattdrayer and @saleem-latif. I commented inline to the comments, if you agree to that, I'll create the other tests and upload them later

mattdrayer commented:
Agreed -- looking forward to the additional tests -- I think we should be good to go after they have been included.

felipemontoya commented:
@mattdrayer and @saleem-latif I added the test and pushed them, to later notice that the PR still needed a rebase. I just rebased it, and after the test complete it should be ready to go

mattdrayer commented:
jenkins run bokchoy

mattdrayer commented:
Green build and two thumbs -- going ahead with the merge -- thanks very much for sticking with this one @felipemontoya!
